### PR TITLE
Some refactoring done for SSO hostobject

### PR DESCRIPTION
**Pull request by malinthaprasan**
Following changes are done.

1. Refactoring  jsFunction_getSAMLAuthRequest() method to reduce complexity and removed duplicated pieces code for building passive/signed Auth requests. 
2. Moving some methods (buildNameID, buildNameIDPolicy, processAcsUrl) to Util class.
3. Adding comments for generic Exception catching, remove places of catching generic Exception where possible.

### CODE REVIEW

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java [hunk 1]
```diff
@@ -234,50 +148,16 @@ public static AuthnRequest setSignature(AuthnRequest authnRequest, String signat
             
             Signer.signObjects(signatureList);
             return authnRequest;
-
-        } catch (Exception e) {
-            throw new Exception("Error while signing the SAML Request message", e);
+        } catch (CertificateEncodingException e) {
```
Comment (nuwand) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java]: Please log and throw exceptions.

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java [hunk 2]
```diff
@@ -53,90 +46,25 @@
      * @return AuthnRequest Object
      * @throws Exception error when bootstrapping
      */
-    public AuthnRequest buildAuthenticationRequest(String issuerId, String nameIdPolicy) throws Exception {
+    public AuthnRequest buildAuthenticationRequest(String issuerId, String acsUrl, boolean isPassive,
+            String nameIdPolicy) throws Exception {
```
Comment (malinthaprasan) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java]: Throwing generic exception fixed in #1494 (https://github.com/wso2/carbon-apimgt/commit/a3c032ab9f7593c6e11cdf1e91fa04d5cc3de991)

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java (hunk 1 above)
Comment (malinthaprasan) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java]: Fixed in #1528 (https://github.com/wso2/carbon-apimgt/commit/c6d6cb237bd7654eda952fb2fbd74863e9899cfc)

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java [hunk 3]
```diff
@@ -234,50 +148,16 @@ public static AuthnRequest setSignature(AuthnRequest authnRequest, String signat
             
             Signer.signObjects(signatureList);
             return authnRequest;
-
-        } catch (Exception e) {
-            throw new Exception("Error while signing the SAML Request message", e);
+        } catch (CertificateEncodingException e) {
+            throw new SignatureException("Error getting certificate", e);
+        } catch (MarshallingException e) {
+            throw new SignatureException("Error while marshalling auth request", e);
```
Comment (malinthaprasan) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java]: Logs added in #1528 (c6d6cb2)

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java [hunk 4]
```diff
@@ -234,50 +148,16 @@ public static AuthnRequest setSignature(AuthnRequest authnRequest, String signat
...
+        } catch (SignatureException e) {
+            throw new SignatureException("Error while signing the SAML Request message", e);
```
Comment (malinthaprasan) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java]: Logs added in #1528 (c6d6cb2)

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java [hunk 5]
```diff
@@ -234,50 +148,16 @@ public static AuthnRequest setSignature(AuthnRequest authnRequest, String signat
...
+        } catch (Exception e) { //buildXMLObject() throws a generic Exception
+            throw new SignatureException("Error while signing the SAML Request message", e);
```
Comment (malinthaprasan) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/builder/AuthReqBuilder.java]: Logs added in #1528 (c6d6cb2)

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/util/Util.java [hunk 6]
```diff
@@ -286,4 +294,59 @@ public static RealmService getRealmService() {
...
+        return nameIdObj;
+    }
+
+    /**
+     * Replaces the ${} in url with system properties and returns
+     *
+     * @param acsUrl
+     * @return
+     */
+    public static String processAcsUrl(String acsUrl){
```
Comment (kasunbg) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/util/Util.java]: Code formatting issue.

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/SAMLSSORelyingPartyObject.java [hunk 7]
```diff
@@ -403,104 +403,31 @@ public static String jsFunction_getUUID(Context cx, Scriptable thisObj,
...
-    public static String jsFunction_getSAMLPassiveAuthRequest(Context cx, Scriptable thisObj,
-                                                              Object[] args,
-                                                              Function funObj) throws Exception {
-        SAMLSSORelyingPartyObject relyingPartyObject = (SAMLSSORelyingPartyObject) thisObj;
-        String nameIdFormat = relyingPartyObject.getSSOProperty(SSOConstants.NAME_ID_POLICY);
 
-        if ((Boolean) args[0])  {
-            String acsUrl = (String) args[1];
+        //check if request signing is required
+        if (!Boolean.valueOf(relyingPartyObject.getSSOProperty(SSOConstants.SIGN_REQUESTS))) {
```
Comment (nuwanw) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/SAMLSSORelyingPartyObject.java]: why evaluating !(not) in "if' even having the "else" block

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/util/Util.java (hunk 6 above)
Comment (malinthaprasan) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/util/Util.java]: Fixed in #1581 (https://github.com/wso2/carbon-apimgt/commit/0e5a93cca133738cc74a26863424e20db485892d)

File: components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/SAMLSSORelyingPartyObject.java (hunk 7 above)
Comment (malinthaprasan) [inline components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/SAMLSSORelyingPartyObject.java]: Fixed in #1581 (https://github.com/wso2/carbon-apimgt/commit/0e5a93cca133738cc74a26863424e20db485892d)

**Pull request merged**